### PR TITLE
Feature/gh 906 handle incoming requests

### LINKED ISSUE #899: WC Flow Coordinator
Create a Flow coordinator to handle transitions between WC screens.
(-) Onboarding
(-) Scanning
(+) Sessions list
(-) Send_Review
- Sign_Review [Next iteration]

### LINKED ISSUE #906: WC Handle incoming messages
Open a review screen with a payload

### LINKED ISSUE #909: WCIncomingRequestRepository and model types
```Swift
public struct WCIncomingRequest {

    public var sessionID: WCSessionID
    public var requestID: String
    public var payload: Payload

    enum Payload {
        case transaction(TransctionID)
    }

}

public protocol WCIncomingRequestRepository {

    func add(_ request: WCIncomingRequest)
    func remove(_ request: WCIncomingRequest)
    func all() -> [WCIncomingRequest]

}

public class InMemoryWCIncomingRequestRepository: WCIncomingRequestRepository
```

### PR DESCRIPTION
**Pull request by sche**
closes #906 
closes #899
closes #909

### CODE REVIEW

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+final class WCPendingTransactionsStore {
+
+    private var pendingTransactions = [WCPendingTransaction]()
+
+    func addPendingTransaction(_ transaction: WCPendingTransaction) {
+        pendingTransactions.append(transaction)
+    }
+
+    func popPendingTransactions() -> [WCPendingTransaction] {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: `removeAll()` is a better name.

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        self.completion = completion
+    }
+
+}
+
+final class WCPendingTransactionsStore {
+
+    private var pendingTransactions = [WCPendingTransaction]()
+
+    func addPendingTransaction(_ transaction: WCPendingTransaction) {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: I think, `add()` would be fine for the name

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+         sessionData: WCSessionData,
+         completion: @escaping (Result<String, Error>) -> Void) {
+        self.transactionID = transactionID
+        self.sessionData = sessionData
+        self.completion = completion
+    }
+
+}
+
+final class WCPendingTransactionsStore {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: The code will be more consistent with existing naming: `...Store` -> `...Repository`

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//
+
+import Foundation
+import MultisigWalletDomainModel
+
+public struct WCPendingTransaction {
+
+    public var transactionID: TransactionID
+    public var sessionData: WCSessionData
+    public var completion: (Result<String, Error>) -> Void
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: Please add comment to what the completion block expects to receive as a result, and when should it be called.

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 5]
```diff
@@ -576,6 +576,11 @@ public class WalletApplicationService: Assertable {
                                processed: tx.processedDate)
     }
 
+    public func transactionHash(_ id: TransactionID) -> String? {
+        guard let tx = DomainRegistry.transactionRepository.find(id: id) else { return nil }
+        return tx.transactionHash?.value
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]:
```suggestion
        return DomainRegistry.transactionRepository.find(id: id)?.transactionHash?.value
```

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 6]
```diff
@@ -825,27 +830,45 @@ public class WalletApplicationService: Assertable {
         return transaction.id.id
     }
 
+    public func draftTransaction(wallet: Wallet, sendTransactionData data: SendTransactionRequiredData) -> TransactionID {
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: `createDraftTransaction(in wallet: Wallet, from data: SendTransactionRequiredData)`

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift [hunk 7]
```diff
@@ -825,27 +830,45 @@ public class WalletApplicationService: Assertable {
...
+        let transactionID = DomainRegistry.transactionService
+            .newDraftTransaction(in: wallet, token: tokenAddress(toAddress: data.to, data: data.data))
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:835]: maybe this or another variant of the `newDraftTransaction` method could return the transaction object directly? the next line queries it from the database.

Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift:835]: Moreover, all of these transaction-related methods could move to the `DomainRegistry.transactionService`

File: MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift [hunk 8]
```diff
@@ -9,24 +9,25 @@ import Common
 public class FailedToConnectSession: DomainEvent {}
 public class SessionUpdated: DomainEvent {}
 public class SendTransactionRequested: DomainEvent {}
+public class NonceUpdated: DomainEvent {}
 
-public typealias WCPendingTransaction = (request: WCSendTransactionRequest, completion: (Result<String, Error>) -> Void)
+extension WCSendTransactionRequest: SendTransactionRequiredData {}
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift:14]: this application service became a stateful object, which is contrary to the expected role of the application service - it should be a facade (just a thin layer in-between) the UI and the domain model. All of the state (persisted or not) is accessed through domain services or domain repositories.

I'll put to my todo list as 'needs refactoring to resolve this design issue.'

File: MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift [hunk 9]
```diff
@@ -123,7 +129,15 @@ extension WalletConnectApplicationService: WalletConnectDomainServiceDelegate {
 
     public func handleSendTransactionRequest(_ request: WCSendTransactionRequest,
                                              completion: @escaping (Result<String, Error>) -> Void) {
-        pendingTransactions.append((request: request, completion: completion))
+        guard let wcSession = sessionRepo.all().first(where: { $0.url == request.url }) else { return }
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift]: session repository needs a new method `find(url: URL) -> WCSession?`

File: MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift [hunk 10]
```diff
@@ -123,7 +129,15 @@ extension WalletConnectApplicationService: WalletConnectDomainServiceDelegate {
...
+        let wallet = walletRepository.selectedWallet()!
+        let txID = walletService.draftTransaction(wallet: wallet, sendTransactionData: request)
+        let sessionData = WCSessionData(wcSession: wcSession)
+        let transaction = WCPendingTransaction(transactionID: txID, sessionData: sessionData) { [unowned self] in
+            completion($0)
+            self.eventPublisher.publish(NonceUpdated())
```
Comment (DmitryBespalov) [inline MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift]: more readable version of the event name would be `NewPendingTransactionCreated()`

File: SafeUIKit/SafeAppUI/MainFlow/MainFlowCoordinator.swift [hunk 11]
```diff
@@ -128,15 +129,32 @@ open class MainFlowCoordinator: FlowCoordinator {
...
+                    self.handleIncomingBETransaction(transactionID)
                 }
             }
         }
     }
 
+    private func handleIncomingBETransaction(_ transactionID: String) {
+        let coordinator = incomingTransactionsManager.coordinator(for: transactionID, source: .browserExtension)
+        enterTransactionFlow(coordinator) { [unowned self] in
+            self.incomingTransactionsManager.releaseCoordinator(by: coordinator.transactionID)
```
Comment (DmitryBespalov) [inline SafeUIKit/SafeAppUI/MainFlow/MainFlowCoordinator.swift:141]: note that this may not be called if the 'back' button was tapped.

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]
👍 

renamings and minor changes needed

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift (hunk 1 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: removeAll does not assume returning some value imo

Comment (DmitryBespalov) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: `popAll()`

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift (hunk 4 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift (hunk 3 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift (hunk 2 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift (hunk 1 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WCPendingTransactionsStore.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift (hunk 5 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift (hunk 6 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WalletApplicationService.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift (hunk 9 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift]: ok

File: MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift (hunk 10 above)
Comment (sche) [reply MultisigWallet/MultisigWalletApplication/WalletConnectApplicationService.swift]: removed

File: SafeUIKit/SafeAppUI/MainFlow/MainFlowCoordinator.swift (hunk 11 above)
Comment (sche) [reply SafeUIKit/SafeAppUI/MainFlow/MainFlowCoordinator.swift:141]: done

**Review by DmitryBespalov — APPROVED** [review]

**Review by DmitryBespalov — APPROVED** [review]

**Pull request merged**